yjit/src/codegen.rs

```diff
@@ -2392,7 +2393,20 @@ fn gen_setinstancevariable(
             asm.cmp(write_val, Qnil.into());
             asm.jbe(skip_wb);
 
+            // If the receiver is old, we need to run the WB
+            let flags_opnd = asm.load(Opnd::mem(64, recv, RUBY_OFFSET_RBASIC_FLAGS));
+            let anded = asm.and(flags_opnd, ((RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1) as u64).into());
+            asm.cmp(anded, ((RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1) as u64).into());
+            asm.je(run_wb);
+
+            // If the GC is doing incremental marking, we need to run the WB
+            let global_ptr = asm.load(unsafe { &rb_yjit_should_run_wb as *const _ as usize }.into());
+            let must_wb_opnd = asm.load(Opnd::mem(8, global_ptr, 0));
+            asm.cmp(must_wb_opnd, 0.into());
+            asm.je(skip_wb);
+
             asm.comment("write barrier");
+            asm.write_label(run_wb);
             asm.ccall(
                 rb_gc_writebarrier as *const u8,
                 vec![
```

Comment:
I think with some small modifications, we may be able to improve the code here to minimize branch mispredictions.

I think our assumption is that most of the time, we don't actually need a write barrier, because most of the time, we're writing to a young object.

By default, the CPU will predict that forward branches are not taken.

It might be possible to structure the code roughly like this:

```
// If the receiver is old, we need to run the WB
let flags_opnd = asm.load(Opnd::mem(64, recv, RUBY_OFFSET_RBASIC_FLAGS));
let anded = asm.and(flags_opnd, ((RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1) as u64).into());
asm.cmp(anded, ((RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1) as u64).into());
asm.je(run_wb); // Unlikely, CPU will predict not taken by default because it's a forward branch

// If the GC is doing incremental marking, we need to run the WB
let global_ptr = asm.load(unsafe { &rb_yjit_should_run_wb as *const _ as usize }.into());
let must_wb_opnd = asm.load(Opnd::mem(8, global_ptr, 0));
asm.cmp(must_wb_opnd, 0.into());
asm.jne(run_wb); // Unlikely, CPU will predict not taken by default because it's a forward branch

asm.jmp(skip_wb) // Unconditional branch, we skip the WB by default

asm.comment("write barrier");
asm.write_label(run_wb);
```

Reply:
I'll try to reorder the code more like this